hrms/overrides/employee_advance.py

```diff
@@ -0,0 +1,230 @@
+            self.db_set("status", status)
+        else:
+            self.status = status
+
+    def after_insert(self):
+        if self.status == "Draft":
+            user_permission_doc = frappe.get_doc(
+                {
+                    "doctype": "User Permission",
+                    "user": self.approver_1,
+                    "allow": "Employee Advance",
+                    "for_value": self.name,
+                    "apply_to_all_doctypes": 1,
+                }
+            )
+            doc = user_permission_doc.insert()
+
+    def on_cancel(self):
+        self.status = "Cancelled"
+        self.docstatus = 2
+
+
+@frappe.whitelist()
+@handle_exceptions_with_readable_message
+def next_state(doc_name):
+    employee_advance_doc = frappe.get_doc("Employee Advance", doc_name)
+
+    if employee_advance_doc.status == EmployeeAdvanceConstants.DRAFT:
+        if frappe.session.user != employee_advance_doc.owner:
+            frappe.throw(_("Only Owner Approve in draft state"))
+        CustomTask.create_link_doc_task_if_all_parents_closed(
+            None,
+            TaskType.APPROVE_EMPLOYEE_ADVANCE,
+            "Employee Advance",
+            employee_advance_doc.name,
+            None,
+            None,
+            [employee_advance_doc.approver_1],
+        )
+        employee_advance_doc.status = EmployeeAdvanceConstants.PENDING_APPROVAL
+        employee_advance_doc.save()
+    elif employee_advance_doc.status == EmployeeAdvanceConstants.PENDING_APPROVAL:
+        if (
+            frappe.session.user != employee_advance_doc.approver_1
+            and frappe.session.user != "Administrator"
+        ):
+            frappe.throw(_("Only the Added approver can approve"))
+
+        employee_advance_doc.status = EmployeeAdvanceConstants.PENDING_APPROVAL_BY_ADMIN
+        CustomTask.mark_tasks_as_completed(
+            "Employee Advance", employee_advance_doc.name, TaskType.APPROVE_EMPLOYEE_ADVANCE
+        )
+        employee_advance_doc.save()
+        employee_advance_doc.submit()
+    elif employee_advance_doc.status == EmployeeAdvanceConstants.PENDING_APPROVAL_BY_ADMIN:
+        if not user_has_role(frappe.session.user, RoleEnum.OFFICE_ADMIN_ROLE.value):
+            frappe.throw(_("Only user having the role of Expense Admin can approve the document"))
+        employee_advance_doc.status = EmployeeAdvanceConstants.PENDING_APPROVAL_BY_HR
+        employee_advance_doc.save()
+    elif employee_advance_doc.status == EmployeeAdvanceConstants.PENDING_APPROVAL_BY_HR:
+        if not user_has_role(frappe.session.user, RoleEnum.HR_ROLE.value):
+            frappe.throw(_("Only user having HR User can approve the document"))
+        employee_advance_doc.status = EmployeeAdvanceConstants.PENDING_PAYMENT
+        employee_advance_doc.save()
+
+
+@frappe.whitelist()
+@handle_exceptions_with_readable_message
+def cancel_doc(doc_name):
+    employee_advance_doc = frappe.get_doc("Employee Advance", doc_name)
```

Comment:
whitelist on which status it is being called

Reply:
Done